docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,165 @@
+- Moved into a “Progressing” state.
+- Moved out of a “Progressing” state and into a “Healthy” state.
+
+
+#### Use case 2:
+As a user, I would like to continue to use the current simultaneous Application update behavior of the ApplicationSet controller.
+
+If no strategy is provided, we propose defaulting to an `AllAtOnce` strategy, which maintains the current behavior.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Initial ApplicationSet Creation
+Application resource creation from an ApplicationSet with a defined strategy looks much like the update process. When a brand new ApplicationSet is first created with a rollout strategy specified, the desired Application resource metadata labels are used to determine when each Application resource is created. Each Application created will be created in the order defined by the steps, if any, and advance to the next step only when a step completes successfully. The same applies if an ApplicationSet is modified to target a different set of destination clusters or namespaces, Applications are created or updated in the order defined by their desired state and the defined step order in the strategy.
+
+#### ApplicationSet Rollout Failure
+In the event that an ApplicationSet spec or template is modified and a target Application resource fails to “complete” a sync in any of the steps, the ApplicationSet rollout is stalled. The ApplicationSet resource will ensure the status field for “ApplicationSetUpToDate” is False. If the maxUpdate allows it, the ApplicationSet will continue updating Applications in the current step, but otherwise, no further changes will be propagated to Application resources by the ApplicationSet, and no steps will advance until each Application can successfully complete a sync. If the ApplicationSet is modified while still in the midst of an ApplicationSet rollout, stalled or otherwise, then the existing rollout is abandoned, the application resources are left in their present state, and the new rollout begins.
+
+#### "Pausing" Application Changes During Rollout
+To implement the “paused” functionality of Applications that are not yet ready to be updated, we have a few options.
+* Disable auto-sync.
+** Potentially conflicts with user provided auto-sync settings.
+** Provides the benefit of being able to see the full diff of the ApplicationSet change.
+* “Pause” the Application.
+** Not Yet Implemented: https://github.com/argoproj/argo-cd/issues/4808
+* Prevent any updates at all to the live Applications via the rolling update strategy defined.
+** This is likely the initial implementation method we'll target.
+
+#### Draft Pull Request
+This PR is now functional and ready for comment. We are actively working on unit tests and documentation.
+https://github.com/wmgroot/argo-cd/pull/1
+
+### Security Considerations
+We do not believe this proposal results in any new security considerations for the ApplicationSet controller.
+
+### Risks and Mitigations
```

Comment:
I would add that maintenance cost is an additional risk. Your work on this has been extensive and excellent. I worry the devs able to dedicate (nearly) full-time effort to Argo CD might not be able to sustain the work with the same level of quality, at least unless their (our) companies adopt this feature for internal use cases and therefore can justify dedicating additional time to it.

I think one way to mitigate that risk would be to make this an opt-in feature and advertise it as "alpha, may be removed." If lots of folks find it useful and adopt it, I have no doubt the community and core contributors will be able to maintain it. If adoption is low, the early adopters will at least have known very well what they were getting into.

Reply:
All that said, I like nifty things, and this is nifty. I favor a gamble on something cool.

Reply:
I'll add the maintenance burden as a risk.

I agree that it's safer to require users to opt-in to enable the feature as an experiment, especially with the potential to solve the same problem through something like an Application dependency DAG.

Would you prefer that the toggle be enabled as an arg or environment variable to the appset controller? I don't see any obvious existing experimental feature flags.
https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L2127-L2133

Reply:
I think an arg + env var would be best. So, the applicationset controller command would have an arg which falls back to an env var (like in [this example](https://github.com/argoproj/argo-cd/blob/a9423aa4f54a0a90db9f858e613845b4912dc02d/cmd/argocd-server/commands/argocd_server.go#L233)). Then the env var would be loaded from the [cmd params configmap](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cmd-params-cm.yaml) via the [controller's deployment manifest](https://github.com/argoproj/argo-cd/blob/master/cmd/argocd-applicationset-controller/commands/applicationset_controller.go).